File: src/maze.py
"""
maze.py — ランダム迷路の生成

再帰的バックトラッキング（DFS）で毎回異なる迷路を作ります。
プレイヤー・川崎・ゴールのスポーン位置もここで決めます。
"""

import random
from typing import NamedTuple

from src.settings import MAZE_HEIGHT, MAZE_WIDTH, TILE_FLOOR, TILE_WALL


class SpawnPositions(NamedTuple):
    """各キャラクターのスポーン座標（グリッド単位）"""
    player: tuple[int, int]
    enemy: tuple[int, int]
    goal: tuple[int, int]


class MazeResult(NamedTuple):
    """迷路生成の結果一式"""
    grid: list[list[int]]
    spawns: SpawnPositions


def generate_maze(
    width: int = MAZE_WIDTH,
    height: int = MAZE_HEIGHT,
) -> MazeResult:
    """
    ランダムな迷路を生成する。

    Args:
        width: グリッド幅（タイル数）
        height: グリッド高さ（タイル数）

    Returns:
        迷路グリッドとスポーン位置
    """
    # 全セルを壁で初期化
    grid = [[TILE_WALL for _ in range(width)] for _ in range(height)]

    def neighbors(cx: int, cy: int) -> list[tuple[int, int]]:
        """2マス先の未訪問セル候補を返す"""
        directions = [(0, -2), (0, 2), (-2, 0), (2, 0)]
        result = []
        for dx, dy in directions:
            nx, ny = cx + dx, cy + dy
            if 1 <= nx < width - 1 and 1 <= ny < height - 1 and grid[ny][nx] == TILE_WALL:
                result.append((nx, ny))
        return result

    # (1,1) から掘り進める
    start_x, start_y = 1, 1
    grid[start_y][start_x] = TILE_FLOOR
    stack = [(start_x, start_y)]

    while stack:
        cx, cy = stack[-1]
        candidates = neighbors(cx, cy)
        if not candidates:
            stack.pop()
            continue

        nx, ny = random.choice(candidates)
        # 現在セルと新セルの間の壁も通路にする
        wall_x = cx + (nx - cx) // 2
        wall_y = cy + (ny - cy) // 2
        grid[wall_y][wall_x] = TILE_FLOOR
        grid[ny][nx] = TILE_FLOOR
        stack.append((nx, ny))

    spawns = _pick_spawn_positions(grid, width, height)
    return MazeResult(grid=grid, spawns=spawns)


def _pick_spawn_positions(
    grid: list[list[int]],
    width: int,
    height: int,
) -> SpawnPositions:
    """
    通路タイルの中からプレイヤー・敵・ゴールの位置を選ぶ。

    プレイヤーは入口付近、敵は対角付近、ゴールはプレイヤーから最も遠い通路。
    """
    floors = [(x, y) for y in range(height) for x in range(width) if grid[y][x] == TILE_FLOOR]
    if len(floors) < 3:
        # フォールバック（通常は起きない）
        return SpawnPositions(player=(1, 1), enemy=(width - 2, height - 2), goal=(width // 2, height // 2))

    player_pos = floors[0]

    # プレイヤーからマンハッタン距離が最大のタイルをゴールに
    goal_pos = max(floors, key=lambda p: abs(p[0] - player_pos[0]) + abs(p[1] - player_pos[1]))

    # プレイヤー・ゴール以外でプレイヤーから最も遠いタイルを敵に
    remaining = [p for p in floors if p != player_pos and p != goal_pos]
    if remaining:
        enemy_pos = max(remaining, key=lambda p: abs(p[0] - player_pos[0]) + abs(p[1] - player_pos[1]))
    else:
        enemy_pos = goal_pos

    return SpawnPositions(player=player_pos, enemy=enemy_pos, goal=goal_pos)

File: web/src/lib/game/maze.ts
import {
  MAZE_HEIGHT,
  MAZE_WIDTH,
  TILE_FLOOR,
  TILE_WALL,
} from "./settings";

export interface SpawnPositions {
  player: [number, number];
  enemy: [number, number];
  goal: [number, number];
}

export interface MazeResult {
  grid: number[][];
  spawns: SpawnPositions;
}

export function generateMaze(
  width: number = MAZE_WIDTH,
  height: number = MAZE_HEIGHT,
): MazeResult {
  const grid: number[][] = Array.from({ length: height }, () =>
    Array(width).fill(TILE_WALL),
  );

  function neighbors(cx: number, cy: number): [number, number][] {
    const directions: [number, number][] = [
      [0, -2],
      [0, 2],
      [-2, 0],
      [2, 0],
    ];
    const result: [number, number][] = [];
    for (const [dx, dy] of directions) {
      const nx = cx + dx;
      const ny = cy + dy;
      if (
        nx >= 1 &&
        nx < width - 1 &&
        ny >= 1 &&
        ny < height - 1 &&
        grid[ny][nx] === TILE_WALL
      ) {
        result.push([nx, ny]);
      }
    }
    return result;
  }

  const startX = 1;
  const startY = 1;
  grid[startY][startX] = TILE_FLOOR;
  const stack: [number, number][] = [[startX, startY]];

  while (stack.length > 0) {
    const [cx, cy] = stack[stack.length - 1];
    const candidates = neighbors(cx, cy);
    if (candidates.length === 0) {
      stack.pop();
      continue;
    }

    const [nx, ny] = candidates[Math.floor(Math.random() * candidates.length)];
    const wallX = cx + (nx - cx) / 2;
    const wallY = cy + (ny - cy) / 2;
    grid[wallY][wallX] = TILE_FLOOR;
    grid[ny][nx] = TILE_FLOOR;
    stack.push([nx, ny]);
  }

  const spawns = pickSpawnPositions(grid, width, height);
  return { grid, spawns };
}

function pickSpawnPositions(
  grid: number[][],
  width: number,
  height: number,
): SpawnPositions {
  const floors: [number, number][] = [];
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      if (grid[y][x] === TILE_FLOOR) {
        floors.push([x, y]);
      }
    }
  }

  if (floors.length < 3) {
    return {
      player: [1, 1],
      enemy: [width - 2, height - 2],
      goal: [Math.floor(width / 2), Math.floor(height / 2)],
    };
  }

  const playerPos = floors[0];
  const manhattan = (a: [number, number], b: [number, number]) =>
    Math.abs(a[0] - b[0]) + Math.abs(a[1] - b[1]);

  const goalPos = floors.reduce((best, p) =>
    manhattan(p, playerPos) > manhattan(best, playerPos) ? p : best,
  );

  const remaining = floors.filter(
    (p) => p[0] !== playerPos[0] || p[1] !== playerPos[1],
  ).filter((p) => p[0] !== goalPos[0] || p[1] !== goalPos[1]);

  const enemyPos =
    remaining.length > 0
      ? remaining.reduce((best, p) =>
          manhattan(p, playerPos) > manhattan(best, playerPos) ? p : best,
        )
      : goalPos;

  return { player: playerPos, enemy: enemyPos, goal: goalPos };
}
